services/scheduler/scheduler.go

```diff
@@ -0,0 +1,329 @@
+)
+
+// Service is responsive for executing tasks and storing them to DB.
+type Service struct {
+	db        *reform.DB
+	scheduler *gocron.Scheduler
+	l         *logrus.Entry
+	tasks     map[string]context.CancelFunc
+	taskMx    sync.RWMutex
+	jobs      map[string]*gocron.Job
+	jobsMx    sync.RWMutex
+	mx        sync.Mutex
+}
+
+// New creates new scheduler service.
+func New(db *reform.DB) *Service {
+	scheduler := gocron.NewScheduler(time.UTC)
+	scheduler.TagsUnique()
+	scheduler.WaitForScheduleAll()
+	return &Service{
+		db:        db,
+		scheduler: scheduler,
+		tasks:     make(map[string]context.CancelFunc),
+		l:         logrus.WithField("component", "scheduler"),
+	}
+}
+
+// Run loads tasks from DB and starts scheduler.
+func (s *Service) Run(ctx context.Context) {
+	if err := s.loadFromDB(); err != nil {
+		s.l.Warn(err)
+	}
+	s.scheduler.StartAsync()
+	<-ctx.Done()
+	s.scheduler.Stop()
+}
+
+// AddParams contains parameters for adding new add to service.
+type AddParams struct {
+	CronExpression string
+	Disabled       bool
+	StartAt        time.Time
+}
+
+// Add adds task to scheduler and save it to DB.
+func (s *Service) Add(task Task, params AddParams) (*models.ScheduledTask, error) {
+	var scheduledTask *models.ScheduledTask
+	var err error
+
+	err = s.db.InTransaction(func(tx *reform.TX) error {
+		scheduledTask, err = models.CreateScheduledTask(tx.Querier, models.CreateScheduledTaskParams{
+			CronExpression: params.CronExpression,
+			StartAt:        params.StartAt,
+			Type:           task.Type(),
+			Data:           task.Data(),
+			Disabled:       params.Disabled,
+		})
+		if err != nil {
+			return err
+		}
+
+		id := scheduledTask.ID
+		task.SetID(id)
+
+		// Don't add job to scheduler if task is disabled.
+		if scheduledTask.Disabled {
+			return nil
+		}
+
+		fn := s.wrapTask(task, id)
+
+		s.mx.Lock()
+		j := s.scheduler.Cron(params.CronExpression).SingletonMode()
+		if !params.StartAt.IsZero() {
+			j = j.StartAt(params.StartAt)
+		}
+		scheduleJob, err := j.Tag(id).Do(fn)
+		s.mx.Unlock()
+		if err != nil {
+			return err
+		}
+
+		s.jobsMx.Lock()
+		s.jobs[id] = scheduleJob
+		s.jobsMx.Unlock()
+
+		nextRun := scheduleJob.NextRun().UTC()
+		lastRun := scheduleJob.LastRun().UTC()
+		scheduledTask, err = models.ChangeScheduledTask(tx.Querier, id, models.ChangeScheduledTaskParams{
+			NextRun: &nextRun,
+			LastRun: &lastRun,
+		})
+		if err != nil {
+			s.l.WithField("id", id).Errorf("failed to set next run for new created task")
+		}
+
+		return nil
+	})
+	return scheduledTask, err
+}
+
+// Remove stops task specified by id and removes it from DB and scheduler.
+func (s *Service) Remove(id string) error {
+	s.taskMx.RLock()
+	if cancel, ok := s.tasks[id]; ok {
+		cancel()
+	}
+	s.taskMx.RUnlock()
+
+	s.jobsMx.Lock()
+	delete(s.jobs, id)
+	s.jobsMx.Unlock()
+
+	err := s.db.InTransaction(func(tx *reform.TX) error {
+		if err := models.RemoveScheduledTask(tx.Querier, id); err != nil {
+			return err
+		}
+
+		s.mx.Lock()
```

Comment:
let's move this block out of the transaction, you need to remove the task after successful commit of transaction.